elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryConstructor.java

```diff
@@ -0,0 +1,507 @@
+                // TODO: filter predicate should support alias
+                .filter(invocation -> invocation.getAlias().equals(fieldName))
+                .findFirst()
+                .orElse(null);
+
+        if (metric != null) {
+            return metric.getFunctionExpression();
+        } else {
+            ColumnProjection dimension = template.getGroupByDimensions().stream()
+                    .filter(projection -> projection.getAlias().equals(fieldName))
+                    .findFirst()
+                    .orElse(null);
+
+            if (dimension == null) {
+                throw new InvalidPredicateException("Having clause field not found " + fieldName);
+            } else {
+                return dimension.getAlias();
+            }
+        }
+    }
+
+    /**
+     * Construct SELECT statement expression with metrics and dimensions directly using alias from subquery.
+     *
+     * @param template query template with nested subquery
+     * @return <code>SELECT function(metric1) AS alias1, dimension1 AS alias2</code>
+     */
+    private String aliasProject(SQLQueryTemplate template) {
+        List<String> metricProjections = template.getMetrics().stream()
+                .map(invocation -> invocation.getFunctionExpression() + " AS " + invocation.getAlias())
+                .collect(Collectors.toList());
+
+        List<String> dimensionProjections = template.getGroupByDimensions().stream()
+                .map(projection -> projection.getAlias() + " AS " + projection.getAlias())
+                .collect(Collectors.toList());
+
+        String projectionClause = String.join(",", metricProjections);
+
+        if (!dimensionProjections.isEmpty()) {
+            projectionClause = projectionClause + "," + String.join(",", dimensionProjections);
+        }
+
+        return projectionClause;
+    }
+
+    /**
+     * Construct directly projection GROUP BY clause using column reference.
+     *
+     * @param groupByDimensions columns to project out
+     * @param queriedTable queried analytic view
+     * @return <code>GROUP BY tb1.col1, tb2.col2, ...</code>
+     */
+    private String referenceProjectGroupBy(Set<ColumnProjection> groupByDimensions, SQLAnalyticView queriedTable) {
+        return "GROUP BY " + groupByDimensions.stream()
+                .map(dimension -> resolveSQLColumnReference(dimension, queriedTable))
+                .collect(Collectors.joining(", "));
+    }
+
+    /**
+     * Construct HAVING clause filter using physical column references. Metric fields need to be aggregated in HAVING.
+     *
+     * @param predicate a filter predicate in HAVING clause
+     * @param table Elide logical table this query is querying
+     * @param template query template
+     * @return an filter/constraint expression that can be put in HAVING clause
+     */
+    private String referenceProjectHavingFilter(FilterPredicate predicate, Table table, SQLQueryTemplate template) {
```

Comment:
I think the referenceProject names are a bit awkward.   _Project_ is both a noun and a verb (less common) with different meanings.  Also, in SQL, the noun form is Projection (not Project).